wandb/sdk/internal/file_stream.py

```diff
@@ -394,8 +391,28 @@ def _thread_body(self):
                             "uploaded": list(uploaded),
                         },
                     )
+
+                    # If we encountered an error trying to publish the
+                    # list of uploaded files, don't reset the `uploaded`
+                    # list. Retry publishing the list on the next heartbeat.
+                    if not isinstance(upload_resp, Exception):
+                        uploaded = set()
+
+            # If there aren't ready chunks or uploaded files, we still want to
+            # send regular heartbeats so the backend doesn't erroneously mark this
+            # run as crashed.
+            if cur_time - posted_data_time > self.heartbeat_seconds:
+                request_with_retry(
```

Comment:
related to @minyoung's comment, i think we may need to update `posted_data_time` or another interval timestamp in this block to ensure that we dont send a heartbeat at every iteration of the while loop after `cur_time - posted_data_time > self.heartbeat_seconds` becomes true; that would result in a lot of heartbeats and may lead to rate limit issues

Reply:
Good catch -- I really mangled this logic.

I think I need to restructure things, because I think if I update `posted_data_time` then because `heartbeat_seconds` is usually lower than `rate_limit_seconds()` then the `heartbeat` can end up starving the uploader, which is what ended up happening originally.

Reply:
OK, updated the logic and hopefully simplified things.

The original structure was good in that we wanted `heartbeats` to take a backseat to chunk uploads, but it had the side-effect of delaying upload notifications.

I decided to instead send upload notifications on both chunk uploads and heartbeats, by modifying `_send` to accept the uploaded set.